libr/core/linux_heap_glibc.c

```diff
@@ -60,6 +60,32 @@ static GHT GH(get_va_symbol)(RCore *core, const char *path, const char *sym_name
 	return vaddr;
 }
 
+static GHT GH(align_address_to_size)(ut64 addr, ut64 align) {
+	return addr + ((align - (addr % align)) % align);
+}
+
+static GHT GH(get_main_arena_with_symbol)(RCore *core, RDebugMap *map) {
+	GHT libc_addr = map->addr;
+	r_return_val_if_fail (libc_addr != GHT_MAX, GHT_MAX);
+
+	GHT main_arena = GHT_MAX;
+	GHT vaddr = GHT_MAX;
+	char *path = strdup (map->name);
+	if (path && r_file_exists (path)) {
+		vaddr = GH (get_va_symbol) (core, path, "main_arena");
+		if (vaddr != GHT_MAX) {
+			main_arena = libc_addr + vaddr;
+		} else {
+			vaddr = GH (get_va_symbol) (core, path, "__malloc_hook");
+			if (vaddr != GHT_MAX) {
+				main_arena = GH (align_address_to_size) (vaddr + libc_addr + sizeof (GHT), 0x20);
```

Comment:
why the `+ sizeof (GHT)`?

Reply:
in memory they are:

```
malloc hook
[optional padding of 8 or 16 bytes]
main_arena
```

So i need to add to the malloc hook address the size of a pointer to make the align works.

My code is based on gef works: https://github.com/hugsy/gef/blob/1d5cffd459bc25e0b02657d0300cb51609b8df6c/gef.py#L596

I still need to do the checks for different cpu architecture and test them.

Reply:
Doesn't this depend a lot on the version of glibc, how it is compiled, etc.? It seems this may be too error prone. Have you tested it on more than one linux distro?

Reply:
I'm testing now this on x86-64 arch linux glibc 2.31 with full debug symbols, ubuntu 20 glibc 2.31, ubuntu 18 glibc 2.27 and it's working.

I will try in the next days with ubuntu 19 (glibc 2.29 I guess) and glibc 2.25/2.26.

Actually the bruteforce method fails sometimes so I hope that this is a valid alternative

Reply:
ok then it's good enough i guess :) Thanks!